### DISCUSSION TITLE
[Proposal] Extend `CallerArgumentExpression` to optionally evaluate constants

### DISCUSSION THREAD

**Discussion by hwoodiwiss** [dotnet/csharplang#5770] [Language Ideas]
# Extend `CallerArgumentExpression` to optionally evaluate constants
- [x] Proposed
- [ ] Prototype: Started - [Roslyn](https://github.com/hwoodiwiss/roslyn/tree/feature/caller-argument-expression-resolve-constants), [Runtime](https://github.com/hwoodiwiss/dotnet-runtime/tree/feature/caller-argument-expression-resolve-constants)
- [ ] Implementation: Not Started
- [ ] Specification: Not Started

## Summary
[summary]: #summary

An extension to the `CallerArgumentExpressionAttribute` that allows a user to specify whether constant values should be evaluated or not.

## Motivation
[motivation]: #motivation

The specific use case that lead me to think of this was for building testing libraries, for providing improved result/assertion outputs.

Given the example:

```c#
	public class Test
	{
		private IDependency _fakeDependency;
		private Service _sut;

		[SetUp]
		public void Setup()
		{
			_fakeDependency = A.Fake<IDependency>();
			_sut = new Service(_fakeDependency);
		}

		[Test]
		public void Test1()
		{
			const string expectedStringValue = "A_CONSTANT_STRING";

			_sut.Serve();

			A.CallTo(() =>
				_fakeDependency.DoDependencyThing(A<ParamObject>
						.That
						.Matches(param => param.Name == expectedStringValue)))
				.MustHaveHappened();
		}
	}
```
The current functionality of `CallerArgumentExpressionAttribute` would allow for the content of the match to be displayed to the user as a message upon the example assertion failing, something like:
`A call to Fake DoDependencyThing matching the condition: param => param.Name == expectedStringValue could not be found`

However, I think it would provide value to be able to display the expected value resolved in-place when populating the `CallerArgumentExpression` string, e.g.:
`A call to Fake DoDependencyThing matching the condition: param => param.Name == "A_CONSTANT_STRING" could not be found`

## Detailed design
[design]: #detailed-design

The Roslyn compiler would be updated to look for a second, boolean, parameter for `CallerArgumentExpressionAttribute`.
The parameter would be `false` by default, and would be named `resolveConstants` (or something similar/appropriate).

```c#
// A usage of CallerArgumentExpression as it currently exists
public void CallerArgumentExpressionAsItCurrentlyExists(Func<string, bool> predicate, [CallerArgumentExpression("predicate")] string predicateText = "") { ... }

// A usage of CallerArgumentExpression with resolveConstants = true
public void CallerArgumentExpressionWithResolveConstants(Func<string, bool> predicate, [CallerArgumentExpression("predicate", true)] string predicateText = "") { ... }
```

Alongside this `System.Runtime.CompilerServices.CallerArgumentExpressionAttribute` will need to be updated to take `resolveConstants` as a parameter, and store it.

```c#
namespace System.Runtime.CompilerServices
{
    [AttributeUsage(AttributeTargets.Parameter, AllowMultiple = false, Inherited = false)]
    public sealed class CallerArgumentExpressionAttribute : Attribute
    {
        public CallerArgumentExpressionAttribute(string parameterName)
        {
            ParameterName = parameterName;
            ResolveConstants = false;
        }

        public CallerArgumentExpressionAttribute(string parameterName, bool resolveConstants)
        {
            ParameterName = parameterName;
            ResolveConstants = resolveConstants;
        }

        public string ParameterName { get; }
        public bool ResolveConstants { get; }
    }
}
```

### Edits
- Removed design for usage with named params
- Updated Prototypes to show Roslyn and Runtime updates

**Comment by CyrusNajmabadi** [comment]
This would be extremely non-trivial.  The current logic for CallerArgumentExpression is that we literally grab the *Text* of the argument is refers to and just copy it in.  We'd now need to actually do some form of evaluation of taht string, figuring out what to actually replace.

**Reply by hwoodiwiss** [reply]
Yeah, I've been looking into how this would be implemented and how things are currently done, and it would be quite a bit of effort.

I've got a trivial-case working prototype (basic bodied lambda wit locally declared variables), however, I think my first port of call is duping all the existing tests and adding a case for with resolveConstants as true/existing so that I can verify more extensively.

I decided to create the proposal at this point to gauge the acceptability of this as a feature before I begin putting work in in-earnest.

**Reply by RikkiGibson** [reply]
It seems like the feature could work by using the `ConstantValue` of the expression and converting it to a string representation. That doesn't sound extremely difficult. Was I missing something?

edit: Yes, I was missing something :) It looks like the suggestion is effectively to dig through the expression used by `CallerArgumentExpression` and do constant folding.

**Reply by hwoodiwiss** [reply]
Yeah, in my current prototype, I'm essentially attempting to bind any IdentifierNameSyntax nodes in the expression and find and replace them in the string of the expression, with some logic for avoiding locally defined symbols, and not duplicating work on already found constant symbols.

For getting the actual values I'm using `boundSymbol.ConstantValue.GetValueToDisplay()`, and in testing this seems to have worked pretty well.

An example test I wrote that passes with this (I temp hardcoded to use the constant resolution code):

```c#
        [ConditionalFact(typeof(CoreClrOnly))]
        public void TestFunc_ExpandedConstantsLambdaExpertDifficulty_String()
        {
            string source = @"
using System;
using System.Runtime.CompilerServices;

class Program
{
    public static int Test(Func<string> i, [CallerArgumentExpression(""i"")] string s = ""<default-arg>"")
    {
        string a = i();
        Console.WriteLine($""{a}, {s}"");
        return 1;
    }

    public static void Main()
    {
        const string testString = ""test_val"";
        const int testInt = 1;
        const float testFloat = 1.12345f;
        Test(() => {
                        string outPut = ""2""+ testString + testInt.ToString() + testFloat.ToString();
                        return outPut;
                   });
        Test(() => ""4"" + testString, ""explicit-value"");
    }
}
";
            var compilation = CreateCompilation(source, targetFramework: TargetFramework.NetCoreApp, options: TestOptions.ReleaseExe, parseOptions: TestOptions.Regular10);
            CompileAndVerify(compilation, expectedOutput: @"2test_val11.12345, () => {
                        string outPut = ""2""+ ""test_val"" + 1.ToString() + 1.12345.ToString();
                        return outPut;
                   }
4test_val, explicit-value").VerifyDiagnostics();
        }
```

Although I appreciate that this is pretty basic.

**Reply by hwoodiwiss** [reply]
Here's a slightly updated version of the test that also works with my current implementation:
I previously had issues with var, but the current version of my resolving code seems to handle the general case well.

```c#
        [ConditionalFact(typeof(CoreClrOnly))]
        public void TestFunc_ExpandedConstantsLambdaExpertDifficulty_String()
        {
            string source = @"
using System;
using System.Runtime.CompilerServices;

class Program
{
    public static int Test(Func<string> i, [CallerArgumentExpression(""i"")] string s = ""<default-arg>"")
    {
        string a = i();
        Console.WriteLine($""{a}, {s}"");
        return 1;
    }

    public static void Main()
    {
        const string testString = ""test_val"";
        const int testInt = 1;
        const float testFloat = 1.12345f;
        Test(() => {
                        var outPut = ""2""+ testString + testInt.ToString() + testFloat.ToString();
                        var otherVar = testString + testInt.ToString();
                        return outPut;
                   });
        Test(() => ""4"" + testString, ""explicit-value"");
    }
}
";
            var compilation = CreateCompilation(source, targetFramework: TargetFramework.NetCoreApp, options: TestOptions.ReleaseExe, parseOptions: TestOptions.Regular10);
            CompileAndVerify(compilation, expectedOutput: @"2test_val11.12345, () => {
                        var outPut = ""2""+ ""test_val"" + 1.ToString() + 1.12345.ToString();
                        var otherVar = ""test_val"" + 1.ToString();
                        return outPut;
                   }
4test_val, explicit-value").VerifyDiagnostics();
        }
```

**Reply by hwoodiwiss** [reply]
@CyrusNajmabadi I've implemented this and my prototype works pretty well ([Roslyn](https://github.com/hwoodiwiss/roslyn/tree/feature/caller-argument-expression-resolve-constants), [Runtime](https://github.com/hwoodiwiss/dotnet-runtime/tree/feature/caller-argument-expression-resolve-constants))

One area of concern would be the implementation of the added propperty `CallerArgumentExpressionResolveConstants` for `PEParameterSymbol`, which I haven't tested yet. Other than that, it's working for all the test cases I've thrown at it so far.
